tests/mock/vscode.ts

```diff
@@ -1024,31 +1093,73 @@ export class VSCode {
       await extension.activate();
 
     for (const [name, provider] of this._webviewProviders) {
-      const webview: any = {};
-      webview.asWebviewUri = uri => path.relative(this.context.extensionUri.fsPath, uri.fsPath);
-      const eventEmitter = new EventEmitter<any>();
-      let initializedPage: Page | undefined = undefined;
-      webview.onDidReceiveMessage = eventEmitter.event;
-      webview.cspSource = 'http://localhost/';
-      const pendingMessages: any[] = [];
-      webview.postMessage = (data: any) => {
-        if (!initializedPage) {
-          pendingMessages.push(data);
+      const { webview, pagePromise } = this._createWebviewAndPage();
+      provider?.resolveWebviewView({ webview, onDidChangeVisibility: () => disposable });
+      this.webViews.set(name, await pagePromise);
+    }
+  }
+
+  dispose() {
+    for (const d of this.context.subscriptions)
+      d.dispose();
+  }
+
+  async webViewsByPanelType(viewType: string) {
+    return (await Promise.all(this._webViewsByPanelType.get(viewType) ?? []));
+  }
+
+  async singleWebViewByPanelType(viewType: string): Promise<Page> {
+    const webViews = this._webViewsByPanelType.get(viewType);
+    if (!webViews?.size) {
+      while (true) {
+        const ev = await new Promise<{ viewType: string, webview: Page }>(r => this._webViewsByPanelTypeAddedResolves.push(r));
+        if (ev.viewType === viewType)
+          return ev.webview;
+      }
+    }
+    if (webViews.size === 1)
+      return await [...webViews][0];
+    throw new Error(`Expected 1 webview of panel type ${viewType}, found ${webViews.size}`);
+  }
+
+  private _createWebviewAndPage() {
+    let initializedPage: Page | undefined = undefined;
+    let currHtml = '';
+    const webview: any = {
+      get html() {
+        return currHtml;
+      },
+      set html(newValue: string) {
+        if (newValue === currHtml)
           return;
-        }
-        initializedPage.evaluate((data: any) => {
-          const event = new globalThis.Event('message');
-          (event as any).data = data;
-          globalThis.dispatchEvent(event);
-        }, data).catch(() => {});
-      };
-      provider.resolveWebviewView({ webview, onDidChangeVisibility: () => disposable });
+        currHtml = newValue;
+        initializedPage?.reload().catch(() => {});
```

Comment:
This would have a race for the cases of undefined initializedPage. Are we reassigning html in our code? What scenario are you trying to test here?

Reply:
Before using `playwrightTestServer` I was reassigning html (it first displayed a loading spinner while lauching a new playwright server and then I was injecting a new html with the iframe), but with the new approach it's no longer needed. I'll remove it.